### ISSUE TITLE
Simplify integration with IStringLocalizer and IStringLocalizer<T>

### ISSUE THREAD

**Issue by dotnetjunkie** [simpleinjector/SimpleInjector#718]
Related to: https://github.com/simpleinjector/SimpleInjector/issues/711.

The SimpleInjector.Integration.ServiceCollection integration package contains an `UseLogging` extension method that allows application components to depend on the non-generic Microsoft `ILogger` abstraction.

It would be useful to have a similar `UseLocalization` extension method that allows application components to depend on the non-generic Microsoft `IStringLocalizer` abstraction.

**Comment by dotnetjunkie** [comment]
[feature-718](https://github.com/simpleinjector/SimpleInjector/tree/feature-718) branch created.

**Comment by kwlin** [comment]
I'll try to fix this this issue :)

**Comment by kwlin** [comment]
- [ ] Create private class RootLocalizer
- [ ] Create private class StringLocalizer<> (wrapper class around Microsoft implementation)
- [ ] Add extensions method UseLocalization to wire things up in SimpleInjector (like UseLogging)
- [ ] Test/try out if AddViewLocalization & AddDataAnnotationsLocalization (https://docs.microsoft.com/en-us/aspnet/core/fundamentals/localization?view=aspnetcore-2.2) works as intended with new implemented stuff.
- [ ] Update documentation
- [ ] Add (Unit)tests? 
- [ ] Additional: maybe split up the file SimpleInjectorServiceCollectionExtensions.cs in to multiple files? Like move logging classes to own file and localization classes to own file?

@dotnetjunkie: What do you think about the above todo list for this issue? Am I missing something?

**Comment by dotnetjunkie** [comment]
@kwlin, I think you're right on track here. Few remarks:

* You can give `RootLocalizer` and `StringLocalizer<T>` each their own file.
* You can move `RootLogger` and `Logger<T>` to their own file as well.
* Great if you could create a PR for the documentation that updates [this file](https://github.com/simpleinjector/Documentation/blob/master/source/servicecollectionintegration.rst), but it's okay to do this later; and otherwise, I will do it myself.
* You can skip writing unit tests. The integration projects contain no unit tests. That's because unit testing them typically says little about their correctness, because only real integration tests would give enough guarantees. I always test the integration by hand and this works fine because integration is typically simple. This is obviously quite different with the Core library. Things are complex and interconnected and I wouldn't even dare maintaining S.I. without unit tests for the core library.
* Keep the `SimpleInjectorServiceCollectionExtensions` together. Moving methods to another class is a breaking change (if we do this, we should at least wait till v5) and when you moved all private classes out, the class will be reasonably small.

**Comment by dotnetjunkie** [comment]
Apparantly, the creation of a `RootLocalizer` makes little sense in the context of `IStringLocalizer<T>`. In contrast to logging, it's not possible to get an "untyped" localizer that is not specific to a particular type. It, therefore, makes no sense to create a `RootLocalizer`. Instead, there is unfortunately no other option than letting Simple Injector throw an exception in case someone resolves `IStringLocalizer` directly (i.e. `container.GetInstance<IStringLocalizer>()`). So the best thing to do is throw an exception:

``` c#
options.Container.RegisterConditional(
    typeof(IStringLocalizer),
    c => c.Consumer is null
        ? throw new ActivationException(ExceptionMessage)
        : typeof(IStringLocalizer<>).MakeGenericType(c.Consumer.ImplementationType),
    Lifestyle.Singleton,
    _ => true);
```

Here is a suggestion for the exception message:

> IStringLocalizer is being resolved directly from the container, but this is not supported as string localizers need to be related to a consuming type. Instead, make IStringLocalizer a constructor dependency of the type it is used in.

**Comment by kwlin** [comment]
Sounds like a plan! I'll try to pick it up this weekend.

**Comment by kwlin** [comment]
> ```cs
> options.Container.RegisterConditional(
>     typeof(IStringLocalizer),
>     c => c.Consumer is null
>         ? throw new ActivationException(ExceptionMessage)
>         : typeof(IStringLocalizer<>).MakeGenericType(c.Consumer.ImplementationType),
>     Lifestyle.Singleton,
>     _ => true);
> ```
> 

Does this mean we don't have to use a sub/wrapper class around "Microsoft.Extensions.Localization.StringLocalizer<T>"? (The logger has a sub class to prevent breaking changes from Microsoft).

**Comment by dotnetjunkie** [comment]
No, I'm sorry; that `typeof(IStringLocalizer<>).MakeGenericType` is incorrect. Should be the wrapper type instead. My bad.

**Comment by kwlin** [comment]
I was testing the  UseLocalization() implementation on the sample app of localization [Microsoft site](https://github.com/aspnet/AspNetCore.Docs/tree/master/aspnetcore/fundamentals/localization/sample/Localization) 

It seems there is also a [IHtmlLocalizer](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.localization.ihtmllocalizer?view=aspnetcore-2.0) with same setup ... Should we make another extensions method or should we register the IHtmlLocalizerFactory and IHtmlLocalizer<> in the UseLocalization() extension?

**Comment by dotnetjunkie** [comment]
That's an interesting finding. That type, however, lives in the Microsoft.AspNetCore.Mvc.Localization NuGet, which has a dependency on Microsoft.AspNetCore.Mvc.Razor. this means this interface can't be added as part of the SimpleInjector.Integration.ServiceLocation package, as it can't take a dependency on Razor stuff.

So if we'd implement this, it should not be part of this issue, and we might need to do some more analysis on how this interface is used, before we decide whether it is useful to add an integration with Simple Injector.

**Comment by dotnetjunkie** [comment]
Moved IHtmlLocalizer discussion to #744

**Comment by kwlin** [comment]
I've put the StringLocalizer<T> wrapper class in different file but then the class have to be internal instead of private. Although it will be still hidden for any users of the SimpleInjector.Integration.ServiceCollection.

**Comment by dotnetjunkie** [comment]
Making classes `internal` has about the same effect as making them `private`, so it's perfectly fine for them to be `internal`.

> Do we still want the Logger also in different class?

Feel free to move the `Logger` classes out as well to their own separate files.

**Comment by dotnetjunkie** [comment]
> And in which branch should I change the documents?

You can make a PR on the [4.0 branch](https://github.com/simpleinjector/Documentation/tree/4.0).

What documentation pages to update:

* `source/servicecollectionintegration.rst` as it contains the documentation for the package in which you added the extension method
* `source/aspnetintegration.rst` where you can add a call to `options.UseLocalization()` next to the call to `options.UseLogging()` in the main code example, to show the existence of this new feature.
* `source/generichostintegration.rst` similarly. It contains a call to `options.UseLogging()` and add a bullet point before the "Using Hosted Services" section about localization.

**Comment by dotnetjunkie** [comment]
Released in Simple Injector v4.7. Thanks for @kwlin for implementing this feature.
